### PR TITLE
feat: unify component with html-template

### PR DESCRIPTION
**Pull request by magicmatatjahu**
**Description**

Unify component with with html-template. For more info see https://github.com/asyncapi/shape-up-process/issues/83 and https://github.com/asyncapi/asyncapi-react/issues/265

Changes:
- remove Fiori styling and start using TailwindCSS - theming at the moment won't be possible, we want to add it in the future in simplifier form that currently - see https://github.com/asyncapi/shape-up-process/issues/83 -> Don't support theming section. `@asyncapi/react-component/lib/styles/fiori.css` is deprecated.
- default styles now is in `@asyncapi/react-component/styles/default.min.css` 
- start supporting `esm` and `cjs` modules. Prepare also compatible with browser umd module.
- supporting almost everything from JSON Schema draft-07 except `dependencies` and `true/false` schemas (due to bug in parser https://github.com/asyncapi/parser-js/issues/232)
- supporting from AsyncAPI 2.0.0 everything except:
  - `id`
  - `operationId`
  - `externalDocs` in operation
  - `discriminator` in AsyncAPI JSON Schema
- `schema` prop can be instance of [`AsyncAPIDocument`](https://github.com/asyncapi/parser-js/blob/master/lib/models/asyncapi.js#L25)
- `config` prop is changed:
  - remove `expand` object due to fact that we want to change expanding functionality
  - move `showErrors` to `show.errors`
  - remove `show.schemas`
  - remove `show.sidebar`
  - rename `show.channels` to `show.operations`
  - remove `pushStateBehaviour` due to fact that we want to change expanding functionality

---

I added `feat:` because I don't think that the current changes (although big) are immediately suitable for 1.0.0 or even 1.0.0-alphaX. We should go further from 0.24.0 and only when we stabilize the component we should go with 1.0.0. Also with 1.0.0-alpha there will be problems with semantic-releases (or not (: ?) @fmvilas @derberg Any thoughts mates?

Thanks for @aeworxet for helping me in this journey! 😄 
 
**Related issue(s)**
See https://github.com/asyncapi/shape-up-process/issues/83
See https://github.com/asyncapi/asyncapi-react/issues/265

### CODE REVIEW

**Comment by derberg** [comment]
Please clearly highlight, as a separate section, in the description what changes are introduced with this PR, what API changes, naming, and styling. It must be clear for others to easily spot what has changed, links to css, and other things like that. Even I'm not able to spot it although I reviewed few PRs you did. Also what new things are supported? some new fields from the spec?

In my opinion, the change in the component is huge and we need to do 1.0.0 and if not sure about stability, then 1.0.0-rc.1. It is about time we follow semver and not do endless minor releases 😄 

I know the argument is: that this is what release 0 is for, to change things and break compatibility when needed, etc.

let me quote semver:
> Major version zero is all about rapid development. If you’re changing the API every day you should either still be in version 0.y.z or on a separate development branch working on the next major version.

We should think about users and face the reality that many people use the component in the production, and having such a large breaking change released under just 0.24 is not good. If people do not use `package-lock.json` and have `"@asyncapi/react": "^0.23.0" then they will face some issues 😞 

I would say the quote from semver says that we should not merge if you have doubts about the stability and release RC from the branch 🤔 

@fmvilas we need another voice here as we have different opinions with @magicmatatjahu here

**Comment by magicmatatjahu** [comment]
@derberg I added changes. Thanks!

**Comment by magicmatatjahu** [comment]
> I would say the quote from semver says that we should not merge if you have doubts about the stability and release RC from the branch 🤔

I don't want release 1.0.0 due to additional fact that we will break in next few month styles again, yes again 😅 and probably due to fact that we want have plugins system, also breaks API, probably only `config` prop. Do semantic releases and workflows are prepared for `alpha` versions?

**Comment by derberg** [comment]
@magicmatatjahu please clarify there that `@asyncapi/react-component/lib/styles/fiori.css` is removed, no longer available.

What about the sandbox projects?

> Do semantic releases and workflows are prepared for alpha versions?
- https://github.com/semantic-release/semantic-release/blob/master/docs/support/FAQ.md#can-i-set-the-initial-release-version-of-my-package-to-001
- https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/pre-releases.md#publishing-pre-releases

**Comment by magicmatatjahu** [comment]
I clarified: 😄 

> move styles from @asyncapi/react-component/lib/styles/fiori.css to @asyncapi/react-component/styles/default.min.css

**Comment by derberg** [comment]
@magicmatatjahu yeap, I saw that. Problem I have there is that it is not "moved", default.css is a completely new file and fiori.css is actually removed. Yes, I know you write it 2 lines above. I just want us to be super clear. So maybe extend first sentence, clarify that Fiory styling is out and therefore fiori.css and that new Tailwind css are in a default.css file, something like this 🙏🏼

**Comment by fmvilas** [comment]
Agree with Łukasz in that we should be aiming for v1 soon, so probably 1.0.0-rc1 or similar would be great. Although, not sure it's something to debate in this PR. Just my two cents.

And also please tell me how to review this PR. It's huge and I have no idea where to start. Feel free to leave comments in the PR (as if you were reviewing your own PR) indicating us how to to review it 🙏

File: LICENSE [hunk 1]
```diff
@@ -187,6 +187,7 @@
      identification within third-party archives.
 
   Copyright (c) 2018-2019 SAP SE or an SAP affiliate company. All rights reserved.
+  Copyright (c) 2020-2021 AsyncAPI Initiative. All rights reserved.
```
Comment (fmvilas) [inline LICENSE:190]:
```suggestion
  Copyright (c) 2020-2021 AsyncAPI contributors
```

**Comment by magicmatatjahu** [comment]
@fmvilas Thanks! I think that debating in this PR is very important because, we must decide where to go and won't release it as 0.24.0.

> And also please tell me how to review this PR. It's huge and I have no idea where to start. Feel free to leave comments in the PR (as if you were reviewing your own PR) indicating us how to to review it 🙏

It's a "'final" PR with unification so I don't think that it must be reviewed second time 😂 All changes are from previous PRs (which had been reviewed).

**Comment by derberg** [comment]
@fmvilas 

> Although, not sure it's something to debate in this PR

the debate runs here as it is a PR from a feature branch to master with `feat:` prefix, which means that once we merge, we release next minor. 

I'm also confused not why PR is not from https://github.com/asyncapi/asyncapi-react/tree/unify-component but from a fork 🤔

**Comment by magicmatatjahu** [comment]
@derberg I tried to do this first, but doing a rebase from about 15 commits got me pissed off 😂 

https://www.youtube.com/watch?v=umDr0mPuyQc

**Comment by magicmatatjahu** [comment]
@fmvilas If you want to check changes, you can generate html-template with react-component onaboard.

```text
ag https://gist.githubusercontent.com/derberg/4e419d6ff5870c7c3f5f443e8bd30535/raw/6188903caf77ad14ee32e22b0c86bf6fa244b580/asyncapi-websocket-kraken.yml https://raw.githubusercontent.com/asyncapi/html-template/42de73c19cd8ff79966e8a1214f16b4f11bb421c/asyncapi-html-template-0.23.0.tgz -o ./sample --force-write
```

**Comment by derberg** [comment]
@magicmatatjahu 🤷🏼 we were reviewing changes to the upstream branch, not fork, maybe this is why Fran wants to do another detailed review because PR is not from upstream 🤷🏼

**Comment by magicmatatjahu** [comment]
@derberg These changes are the same. From `unify-component` you will see these same changes (144 files), because at the end you wanna merge to master

**Comment by magicmatatjahu** [comment]
@derberg @fmilas Or maybe we will treat `unify-component` branch as our `prerelease`? then we will have `@asyncapi/react-componentv1.0.0-unify-component-X`?

**Comment by fmvilas** [comment]
Oh sorry, I didn't understand this PR correctly. I'd not go yet for 1.0.0-alpha as the design and structure of the project are going to change substantially soon. For instance, the "React component" will become just a component of a UI library.

**Comment by magicmatatjahu** [comment]
@derberg But how to handle problem with breaking changes, especially styling? At first I thought to go with 0.24.0, but Łukasz's idea is better. You can see in the linked issue from the `backstage` project that people are using the custom styling. If we go with alpha or rc, we can say `ok, we hadn't much time for UI library, so we will treat alpha/rc branch as something like 0.0.1 version with breaking changes in a short time`. What do you think?

**Comment by fmvilas** [comment]
I see. Agree to go with 1.0.0-rc1 or similar then 👍

**Comment by derberg** [comment]
I suggest renaming feature branch to `beta` and release as 1.0.0-beta.1, beta.2 etc

- [here](https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/pre-releases.md#working-on-a-future-release) you have explanation how it works 
- possible configuration [here](https://github.com/semantic-release/semantic-release/blob/master/docs/usage/configuration.md#configuration) but `beta` branch is treated as `prerelease` by default. Of course you can keep current branch name if you want, and mark as `prerelease` but then just add it to `semantic-release` config

Did I tell you how much I love conventional commits and semantic-release? 😄 

cools side effect of this exercise here is that whatever you work on now, and make it work, will apply to AsyncAPI spec releases that we need to configure for 2.1 release. 

What more you will have to do? I think that:
- probably https://github.com/asyncapi/asyncapi-react/blob/master/.github/workflows/if-nodejs-release.yml must point to feature branch too
- might be that you need to change something [here](https://github.com/asyncapi/asyncapi-react/blob/master/.github/workflows/release-wc-and-playground.yml) because, initial beta release will be marked as `@beta` and not `@latest` on npm, so the same should happen with release of the web component

@magicmatatjahu Let me know if you need some help, you are not alone my friend, I'm with you 😄

**Comment by magicmatatjahu** [comment]
@derberg Thanks! Maybe `next` not `beta`? 😄

**Comment by derberg** [comment]
@magicmatatjahu from my point of view, the most important is we do not release it under 0.24 and users are protected. If branch and release name is `beta`, `next`, or `maciek-is-nice`, is not that important for me. Choose whatever you thing will be best

**Comment by magicmatatjahu** [comment]
We decided to go with `1.0.0-next.X` version. See more in [`next`](https://github.com/asyncapi/asyncapi-react/tree/next) branch.

**Pull request closed without merging**